pandas/core/dtypes/cast.py

```diff
@@ -1696,7 +1696,10 @@ def construct_1d_arraylike_from_scalar(
 
     else:
 
-        if length and is_integer_dtype(dtype) and isna(value):
+        if is_integer_dtype(dtype) and isna(value):
+            if not length:
+                # GH 47391: numpy > 1.24 will raise filling np.nan into int dtypes
+                return np.array([], dtype=dtype)
```

Comment:
this seems like a weird case.  how does it happen?  is it clear that we'd want to prioritize the dtype as being "right" instead of the value?

Reply:
I posted an example of the state that's reached in https://github.com/numpy/numpy/issues/21784

Namely before this change, `length=0` would pass all the if checks down to `np.empty(0, dtype=integer).fill(np.nan)` which in numpy < 1.24 would just return `np.array([], dtype=integer)` but in numpy >=1.24 will raise

Reply:
I think the question (also for me from NumPy!) is which *pandas* code runs into this path.  For pandas, the question is what the actual result should be (in the future).  For me the question is how bad it will be if that code path breaks.  Because especially if it is bad, we may want to make sure it doesn't break yet (from within NumPy).

(At this point I suspect that at least the NaN case may need a work-around in NumPy as well.)

Reply:
So an example test where this is hit is 

```
        s = Series([], index=pd.date_range(start="2018-01-01", periods=0), dtype=int)
        result = s.apply(lambda x: x)
        tm.assert_series_equal(result, s)
```

so I _think_ when operating over these empty Series/DataFrames, the value representation is `np.nan` when `construct_1d_arraylike_from_scalar(np.nan, length=0, dtype=dtype)` is called.

`construct_1d_arraylike_from_scalar` has a `if length and is_integer_dtype(dtype) and isna(value)` condition to ensure that integer dtypes were not coerced to float64 (because we relied on `np.empty(0, dtype=integer).fill(np.nan) == np.array([], dtype=integer)`) for these empty Series/DataFrames.

Reply:
So for these empty-like cases in pandas, it appears pandas was relying on `np.empty(0, dtype=integer).full(np.nan)` to preserve integer dtypes. Having pandas explicitly preserve the dtype for these empty cases e.g. `np.array([], dtype=integer)` is an okay change to make on our end IMO.

Reply:
sgtm. (but why not keep this bit unchanged and just put the `subarr.fill(value)` on L1715 inside a `if length:` to skip for all zero-length arrays?)

Reply:
Ah sure I can make the change there instead